src/mechanics/integrands/Base.h

```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include <memory>
+
+namespace NuTo
+{
+namespace Integrand
+{
+class Base
+{
+public:
+    virtual std::unique_ptr<Base> Clone() const = 0;
+    virtual ~Base() = default;
+
+    template <typename TIntegrand>
+    TIntegrand& As()
+    {
```

Comment:
Out of curiosity: Does template argument deduction work here?

Reply:
Nope
~~~cpp
struct Base
{
    virtual ~Base() = default;
    template <typename T>
    T& As()
    {
        return dynamic_cast<T&>(*this);
    }
};
struct Foo : Base {};
int main()
{
    Foo foo;
    Base& base = foo;
    Foo& fooCasted = base.As();
}
~~~

~~~
template_deduction_return.cpp:20:27: error: no matching member function for call to 'As'                                                                                                                          
    Foo& fooCasted = base.As();                                                                                                                                                                                   
                     ~~~~~^~                                                                                                                                                                                      
template_deduction_return.cpp:7:8: note: candidate template ignored: couldn't infer template argument 'T'                                                                                                         
    T& As() 
~~~